File: backend/src/models/tickets.js
const mongoose = require('mongoose');
const crypto = require('crypto');

// Ticket Schema
const ticketSchema = new mongoose.Schema({
  ticket_id: {
    type: String,
    required: true,
    unique: true,
    default: () => require('ulid').ulid()
  },
  tenant_id: {
    type: String,
    required: true,
    ref: 'Tenant'
  },
  store_id: {
    type: String,
    required: true,
    ref: 'Store'
  },
  sale_id: {
    type: String,
    required: true,
    ref: 'Sale'
  },
  package_id: {
    type: String,
    required: true,
    ref: 'Package'
  },
  short_code: {
    type: String,
    required: true,
    unique: true,
    uppercase: true
  },
  qr_token: {
    type: String,
    required: true,
    unique: true
  },
  signature: {
    type: String,
    required: true
  },
  type: {
    type: String,
    required: true,
    enum: ['single', 'multi', 'timepass', 'credit']
  },
  quota_or_minutes: {
    type: Number,
    required: true,
    min: 1
  },
  used: {
    type: Number,
    default: 0,
    min: 0
  },
  valid_from: {
    type: Date,
    required: true
  },
  valid_to: {
    type: Date,
    required: true
  },
  status: {
    type: String,
    enum: ['active', 'cancelled', 'refunded', 'expired'],
    default: 'active'
  },
  lot_no: {
    type: String,
    required: true
  },
  shift_id: {
    type: String,
    ref: 'Shift'
  },
  issued_by: {
    type: String,
    required: true,
    ref: 'Employee'
  },
  price: {
    type: Number,
    required: true,
    min: 0
  },
  payment_method: {
    type: String,
    required: true,
    enum: ['cash', 'qr', 'card', 'other']
  },
  printed_count: {
    type: Number,
    default: 0,
    min: 0
  },
  device_binding: [{
    type: String,
    ref: 'Device'
  }],
  access_zones: [{
    type: String,
    ref: 'AccessZone'
  }],
  created_at: {
    type: Date,
    default: Date.now
  },
  updated_at: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' }
});

// Redemption Schema
const redemptionSchema = new mongoose.Schema({
  redemption_id: {
    type: String,
    required: true,
    unique: true,
    default: () => require('ulid').ulid()
  },
  ticket_id: {
    type: String,
    required: true,
    ref: 'Ticket'
  },
  device_id: {
    type: String,
    required: true,
    ref: 'Device'
  },
  timestamp: {
    type: Date,
    default: Date.now
  },
  result: {
    type: String,
    required: true,
    enum: ['pass', 'fail']
  },
  reason: {
    type: String,
    enum: [
      'expired', 'duplicate_use', 'wrong_device', 'not_started',
      'exhausted', 'invalid_sig', 'cancelled', 'refunded'
    ]
  },
  metadata: {
    remaining_quota: Number,
    remaining_minutes: Number,
    device_location: String,
    scan_location: {
      lat: Number,
      lng: Number
    },
    notes: String
  },
  created_at: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' }
});

// Indexes
ticketSchema.index({ ticket_id: 1 });
ticketSchema.index({ qr_token: 1 });
ticketSchema.index({ short_code: 1 });
ticketSchema.index({ tenant_id: 1, store_id: 1 });
ticketSchema.index({ sale_id: 1 });
ticketSchema.index({ status: 1 });
ticketSchema.index({ valid_from: 1, valid_to: 1 });
ticketSchema.index({ lot_no: 1 });

redemptionSchema.index({ redemption_id: 1 });
redemptionSchema.index({ ticket_id: 1 });
redemptionSchema.index({ device_id: 1 });
redemptionSchema.index({ timestamp: 1 });
redemptionSchema.index({ result: 1 });

// Pre-save middleware for ticket generation
ticketSchema.pre('save', function(next) {
  if (this.isNew) {
    // Generate short code if not provided
    if (!this.short_code) {
      this.short_code = this.generateShortCode();
    }
    
    // Generate QR token if not provided
    if (!this.qr_token) {
      this.qr_token = this.generateQRToken();
    }
    
    // Generate signature if not provided
    if (!this.signature) {
      this.signature = this.generateSignature();
    }
  }
  next();
});

// Methods
ticketSchema.methods.generateShortCode = function() {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let result = '';
  for (let i = 0; i < 7; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
};

ticketSchema.methods.generateQRToken = function() {
  return require('ulid').ulid();
};

ticketSchema.methods.generateSignature = function() {
  const hmacSecret = process.env.HMAC_SECRET_V1 || 'default-secret';
  const data = `${this.ticket_id}:${this.qr_token}:${this.valid_from}:${this.valid_to}`;
  return crypto.createHmac('sha256', hmacSecret).update(data).digest('hex');
};

ticketSchema.methods.verifySignature = function() {
  const hmacSecret = process.env.HMAC_SECRET_V1 || 'default-secret';
  const data = `${this.ticket_id}:${this.qr_token}:${this.valid_from}:${this.valid_to}`;
  const expectedSignature = crypto.createHmac('sha256', hmacSecret).update(data).digest('hex');
  return this.signature === expectedSignature;
};

ticketSchema.methods.isValid = function() {
  const now = new Date();
  return this.status === 'active' && 
         now >= this.valid_from && 
         now <= this.valid_to &&
         this.used < this.quota_or_minutes;
};

ticketSchema.methods.canRedeem = function(deviceId) {
  if (!this.isValid()) return false;
  
  // Check device binding if specified
  if (this.device_binding && this.device_binding.length > 0) {
    return this.device_binding.includes(deviceId);
  }
  
  return true;
};

ticketSchema.methods.getRemainingQuota = function() {
  return Math.max(0, this.quota_or_minutes - this.used);
};

ticketSchema.methods.getRemainingMinutes = function() {
  if (this.type !== 'timepass') return null;
  return Math.max(0, this.quota_or_minutes - this.used);
};

// Virtual for QR payload
ticketSchema.virtual('qr_payload').get(function() {
  return JSON.stringify({
    ticket_id: this.ticket_id,
    qr_token: this.qr_token,
    short_code: this.short_code,
    type: this.type,
    valid_from: this.valid_from.toISOString(),
    valid_to: this.valid_to.toISOString(),
    sig: this.signature
  });
});

// Create models
const Ticket = mongoose.model('Ticket', ticketSchema);
const Redemption = mongoose.model('Redemption', redemptionSchema);

module.exports = {
  Ticket,
  Redemption
};

File: backend-fastapi/app/models/tickets.py
"""
Ticket Models
"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum

from .core import BaseDocument


class TicketType(str, Enum):
    """Ticket types"""
    SINGLE = "single"
    MULTI = "multi"
    TIME_BASED = "time_based"


class TicketStatus(str, Enum):
    """Ticket status"""
    ACTIVE = "active"
    REDEEMED = "redeemed"
    EXPIRED = "expired"
    REFUNDED = "refunded"
    CANCELLED = "cancelled"


class RedemptionResult(str, Enum):
    """Redemption results"""
    PASS = "pass"
    FAIL = "fail"
    ALREADY_USED = "already_used"
    EXPIRED = "expired"
    NOT_STARTED = "not_started"
    QUOTA_EXHAUSTED = "quota_exhausted"
    INVALID_SIGNATURE = "invalid_signature"


class Ticket(BaseDocument):
    """Ticket model"""
    
    ticket_id: str = Field(..., description="Unique ticket identifier")
    tenant_id: str = Field(..., description="Parent tenant ID")
    store_id: str = Field(..., description="Parent store ID")
    sale_id: str = Field(..., description="Parent sale ID")
    package_id: str = Field(..., description="Package ID")
    type: TicketType = Field(..., description="Ticket type")
    quota_or_minutes: int = Field(..., ge=1, description="Quota or minutes")
    used: int = Field(default=0, ge=0, description="Usage count")
    valid_from: datetime = Field(..., description="Valid from date")
    valid_to: datetime = Field(..., description="Valid to date")
    lot_no: str = Field(..., description="Lot number")
    shift_id: str = Field(..., description="Shift ID")
    issued_by: str = Field(..., description="Issuer employee ID")
    price: float = Field(..., ge=0, description="Ticket price")
    payment_method: str = Field(..., description="Payment method")
    qr_token: str = Field(..., description="QR token")
    qr_payload: Dict[str, Any] = Field(..., description="QR payload")
    short_code: str = Field(..., description="Short code")
    signature: str = Field(..., description="HMAC signature")
    status: TicketStatus = Field(default=TicketStatus.ACTIVE, description="Ticket status")
    settings: Dict[str, Any] = Field(default_factory=dict, description="Ticket settings")
    
    class Config:
        collection = "tickets"


class TicketRedeemRequest(BaseModel):
    """Ticket redemption request"""
    
    qr_token: str = Field(..., description="QR token")
    device_id: str = Field(..., description="Redemption device ID")


class TicketRedeemResponse(BaseModel):
    """Ticket redemption response"""
    
    result: RedemptionResult = Field(..., description="Redemption result")
    reason: str = Field(..., description="Result reason")
    remaining: Optional[int] = Field(default=None, description="Remaining uses")
    ticket: Optional[Dict[str, Any]] = Field(default=None, description="Ticket information")


class TicketResponse(BaseModel):
    """Ticket response"""
    
    ticket_id: str = Field(..., description="Ticket ID")
    package_name: str = Field(..., description="Package name")
    type: TicketType = Field(..., description="Ticket type")
    quota_or_minutes: int = Field(..., description="Quota or minutes")
    used: int = Field(..., description="Usage count")
    status: TicketStatus = Field(..., description="Ticket status")
    valid_from: datetime = Field(..., description="Valid from date")
    valid_to: datetime = Field(..., description="Valid to date")
    short_code: str = Field(..., description="Short code")
    qr_payload: Dict[str, Any] = Field(..., description="QR payload")
    redemptions: List[Dict[str, Any]] = Field(default_factory=list, description="Redemption history")


class Redemption(BaseDocument):
    """Redemption model"""
    
    redemption_id: str = Field(..., description="Unique redemption identifier")
    ticket_id: str = Field(..., description="Ticket ID")
    device_id: str = Field(..., description="Redemption device ID")
    employee_id: Optional[str] = Field(default=None, description="Redemption employee ID")
    result: RedemptionResult = Field(..., description="Redemption result")
    reason: str = Field(..., description="Result reason")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Redemption timestamp")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="Redemption metadata")
    
    class Config:
        collection = "redemptions"


class TicketValidation(BaseModel):
    """Ticket validation model"""
    
    qr_token: str = Field(..., description="QR token")
    signature: str = Field(..., description="HMAC signature")
    timestamp: datetime = Field(..., description="Validation timestamp")
    device_id: Optional[str] = Field(default=None, description="Validation device ID")


class TicketGenerationRequest(BaseModel):
    """Ticket generation request"""
    
    sale_id: str = Field(..., description="Sale ID")
    package_id: str = Field(..., description="Package ID")
    quantity: int = Field(..., ge=1, description="Ticket quantity")
    valid_days: int = Field(default=30, ge=1, description="Validity in days")


class TicketGenerationResponse(BaseModel):
    """Ticket generation response"""
    
    tickets: List[TicketResponse] = Field(..., description="Generated tickets")
    count: int = Field(..., description="Generated ticket count")
